src/protocol/sort/shuffle.rs

```diff
@@ -228,88 +213,187 @@ mod tests {
+        use std::iter::zip;
 
-        let shuffled: [_; 3] = {
-            let [ctx0, ctx1, ctx2] = narrow_contexts(&context, &ShuffleOrUnshuffle::Shuffle);
-            let [shares0, shares1, shares2] = shares;
-            let h0_future = shuffle_shares(shares0, (perm1.0.as_slice(), perm1.1.as_slice()), ctx0);
-            let h1_future = shuffle_shares(shares1, (perm2.0.as_slice(), perm2.1.as_slice()), ctx1);
-            let h2_future = shuffle_shares(shares2, (perm3.0.as_slice(), perm3.1.as_slice()), ctx2);
-
-            join3(h0_future, h1_future, h2_future).await
-        };
-        let unshuffled: [_; 3] = {
-            let [ctx0, ctx1, ctx2] = narrow_contexts(&context, &ShuffleOrUnshuffle::Unshuffle);
-            let [shuffled0, shuffled1, shuffled2] = shuffled;
-            let h0_future =
-                unshuffle_shares(shuffled0, (perm1.0.as_slice(), perm1.1.as_slice()), ctx0);
-            let h1_future =
-                unshuffle_shares(shuffled1, (perm2.0.as_slice(), perm2.1.as_slice()), ctx1);
-            let h2_future =
-                unshuffle_shares(shuffled2, (perm3.0.as_slice(), perm3.1.as_slice()), ctx2);
-
-            // When unshuffle and shuffle are called with same step, they undo each other's effect
-            join3(h0_future, h1_future, h2_future).await
+        use crate::ff::Fp31;
+        use crate::protocol::context::Context;
+        use crate::protocol::sort::shuffle::{
+            get_two_of_three_random_permutations, shuffle_shares, unshuffle_shares,
         };
+        use crate::protocol::QueryId;
+        use crate::test_fixture::{Reconstruct, Runner, TestWorld};
+
+        #[tokio::test]
+        async fn malicious() {
+            const BATCHSIZE: u8 = 25;
+            let world = TestWorld::new(QueryId);
+
+            let input: Vec<u8> = (0..BATCHSIZE).collect();
+            let hashed_input: HashSet<u8> = input.clone().into_iter().collect();
+
+            let input_u128: Vec<u128> = input.iter().map(|x| u128::from(*x)).collect();
+
+            let result = world
+                .malicious(
+                    input_u128.clone().into_iter().map(Fp31::from),
+                    |ctx, m_shares| async move {
+                        let perms =
+                            get_two_of_three_random_permutations(BATCHSIZE.into(), ctx.prss_rng());
+                        shuffle_shares(
+                            m_shares,
+                            (perms.0.as_slice(), perms.1.as_slice()),
+                            ctx.clone(),
+                        )
+                        .await
+                        .unwrap()
+                    },
+                )
+                .await;
+
+            let mut hashed_output_secret = HashSet::new();
+            let mut output_secret = Vec::new();
+            for (r0, (r1, r2)) in zip(result[0].iter(), zip(result[1].iter(), result[2].iter())) {
+                let val = (r0, r1, r2).reconstruct();
```

Comment:
You should be able to do:

```suggestion
            for value in result.reconstruct() {
```